File: 274.h-index.py
class Solution:
    def hIndex(self, citations: List[int]) -> int:
        citations.sort(reverse=True)

        for i in range(len(citations)):
            if i+1 > citations[i]:
                return i

        return len(citations)
	
	def hIndex_full(self, citations: List[int]) -> int:
        n = len(citations)
        hist = [0]*(n+1)
        for c in citations:
            if c >= n:
                hist[n]+=1
            else:
                hist[c]+=1

        csum = 0
        for i in range(n,-1,-1):
            csum += hist[i]
            if csum>=i:
                return i
        return 0

File: 274.h-index.cpp
#include <vector>
using namespace std;
class Solution {
public:    
    // Space O(1)
    // Time O(nlogn)
    int hIndex(vector<int>& citations) {
        std::sort(citations.begin(), citations.end(), greater<int>());
        for(int i = 0; i < citations.size(); ++i){
            if(citations[i]<=i)
                return i;
        }
        return citations.size();
    }


    // Space O(n)
    // Time O(n)
    int hIndex_optimize(vector<int>& citations) {
        int n = citations.size();
        int * hist = new int[n+1]();
        
        for(int c: citations){
            if(c >= n)
                ++hist[n];
            else
                ++hist[c];
        }

        int cnt = 0;
        for(int i = n; i >=0; --i){
            cnt += hist[i];
            if(cnt >= i)
                return i;
        }
        return 0;
    }
};